From 8ad657e2482db296aa39ac869119442aa730edc3 Mon Sep 17 00:00:00 2001
From: Chandu Babu N <chandu@amd.com>
Date: Wed, 28 Jun 2017 20:40:08 +0530
Subject: [PATCH 6/7] st/omx: handle invalid timestamps better for frc

Handle buffers with 0 timestamps better by keeping track of the en time of the
previous buffer and assuming the 0 timestamp buffer goes right after the
previous one.
---
 src/gallium/state_trackers/omx/vid_dec.c | 1 +
 src/gallium/state_trackers/omx/vid_enc.c | 4 ++++
 2 files changed, 5 insertions(+)

diff --git a/src/gallium/state_trackers/omx/vid_dec.c b/src/gallium/state_trackers/omx/vid_dec.c
index 313bc0a..e0d3043 100644
--- a/src/gallium/state_trackers/omx/vid_dec.c
+++ b/src/gallium/state_trackers/omx/vid_dec.c
@@ -656,6 +656,7 @@ static void vid_dec_FrameDecoded(OMX_COMPONENTTYPE *comp, OMX_BUFFERHEADERTYPE*
       }
       output->nFilledLen = output->nAllocLen;
       output->nTimeStamp = input->nTimeStamp;
+      output->nTickCount = input->nTickCount;
    }
 
    if (eos && input->pInputPortPrivate)
diff --git a/src/gallium/state_trackers/omx/vid_enc.c b/src/gallium/state_trackers/omx/vid_enc.c
index ec3b281..c80dcd5 100644
--- a/src/gallium/state_trackers/omx/vid_enc.c
+++ b/src/gallium/state_trackers/omx/vid_enc.c
@@ -1332,6 +1332,10 @@ static OMX_ERRORTYPE vid_enc_EncodeFrame(omx_base_PortType *port, OMX_BUFFERHEAD
 
       /* Two frames are available to make a choice */
       prev_ts = prev_task->timestamp;
+      /* get the time of the next expected buffer timestamp, we use this when the
+       * next buffer has 0 as a timestamp */
+      if(!task->timestamp)
+          task->timestamp= prev_ts + (task->tick_count/1000);
       new_ts = task->timestamp;
 
       /* Workaround if decoder does not use discontinuity flags to
-- 
2.7.4

